src/value/mod.rs

```diff
@@ -155,64 +155,232 @@ impl Value {
+            Value::Struct(i) => Ok(i),
+            _ => Err(()),
+        }
+    }
+}
+
 impl From<Vec<u8>> for Value {
     fn from(other: Vec<u8>) -> Self {
         Value::Base64(other)
     }
 }
+
+impl<'a> TryFrom<&'a Value> for &'a Vec<u8> {
+    type Error = ();
+    fn try_from(value: &'a Value) -> Result<Self, Self::Error> {
+        match value {
+            Value::Base64(i) => Ok(i),
+            _ => Err(()),
+        }
+    }
+}
+
+// Due to lack of variadic traits in Rust we have do this kinda hacky support for tuple destructuring
+// See https://stackoverflow.com/questions/38863781/how-to-create-a-tuple-from-a-vector
+// and https://gist.github.com/PoignardAzur/aea33f28e2c58ffe1a93b8f8d3c58667
+
+/// Trait allowing for tuple destructuring iter of Value into a tuple
+pub trait TryCollectValue<'a, T> {
+    fn try_collect_value(&'a mut self) -> Option<T>;
+}
+
+// We define a macro that actually create the implementation of our TryCollect trait for
+// a tuple of a given size.
+macro_rules! impl_try_collect_value_tuple {
+    ($A:ident $($B:ident)*) => {
+        impl<'a, Iter: Iterator<Item = &'a Value>, $A: TryFrom<&'a Value> $(, $B: TryFrom<&'a Value>)*> TryCollectValue<'a, ($A $(, $B)*)> for Iter {
+            fn try_collect_value(&'a mut self) -> Option<($A $(, $B)*)> {
+                let r = ($A::try_from(self.next()?).ok()?
+                $(, $B::try_from(self.next()?).ok()?)*
+                );
+                Some(r)
+            }
+        }
+    }
+}
+
+// We then invoke the macro for each size of tuple we want to support,
+// convention in Rust is to support up to 12
+// This generates a impl of TryCollectValue for Iterator<Value> to covert to each tuple size
+impl_try_collect_value_tuple!(A B);
+impl_try_collect_value_tuple!(A B C);
+impl_try_collect_value_tuple!(A B C D);
+impl_try_collect_value_tuple!(A B C D E);
+impl_try_collect_value_tuple!(A B C D E F);
+impl_try_collect_value_tuple!(A B C D E F G);
+impl_try_collect_value_tuple!(A B C D E F G H);
+impl_try_collect_value_tuple!(A B C D E F G H I);
+impl_try_collect_value_tuple!(A B C D E F G H I J);
+impl_try_collect_value_tuple!(A B C D E F G H I J K);
+impl_try_collect_value_tuple!(A B C D E F G H I J K L);
+
+#[cfg(test)]
+mod test {
+    use super::TryCollectValue;
+    use crate::Value;
+
+    #[test]
+    fn test_tuple_destructure() {
+        let vec = vec![Value::Int(1), Value::String("2".into())];
+
+        let (a, b): (i32, String) = vec.iter().try_collect_value().unwrap();
+        assert_eq!(a, 1);
+        assert_eq!(b, "2");
+    }
+
+    // TODO would be a nice to have to figure this out
+    // Current impl doesn't handle nested vector types...
+    // would need to recursively call try_from...
+    // #[test]
+    // fn test_nested_array_destructure() {
+    //     let vec = vec![vec![vec![Value::String("2".into())]]];
+
+    //     let a : (Vec<Vec<String>>, ) = vec.iter().try_collect_value().unwrap();
+    // }
+}
```

Comment:
Is there any way we could lean on serde to deserialize the Value into the right type?

Reply:
I am so so so very glad you asked this question. I'm still new to serde and tried to tackle this problem with my "classical" rust tools. THIS IS SUCH A BETTER WAY TO DO IT.

Anywho PR should have another looksie, I think it is significantly improved.